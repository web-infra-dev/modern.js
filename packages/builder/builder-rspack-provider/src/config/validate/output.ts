import { sharedOutputConfigSchema, z } from '@modern-js/builder-shared';
import type { OutputConfig } from '../../types';

export const outputConfigSchema: z.ZodType<OutputConfig> =
  sharedOutputConfigSchema
    .extend({
      convertToRem: z.union([z.boolean(), z.instanceof(Object)]),
      externals: z.record(z.string()),
    })
    .partial();
