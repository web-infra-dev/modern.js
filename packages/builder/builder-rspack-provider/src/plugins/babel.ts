import {
  mergeRegex,
  applyScriptCondition,
  JS_REGEX,
  TS_REGEX,
} from '@modern-js/builder-shared';
import { cloneDeep, isEqual } from '@modern-js/utils/lodash';
import { BuilderPlugin, NormalizedConfig } from '../types';
import type { BabelOptions } from '@modern-js/types';
// eslint-disable-next-line @typescript-eslint/no-restricted-imports
import { DEFAULT_BABEL_PRESET_TYPESCRIPT_OPTIONS } from '@modern-js/utils';

export const builderPluginBabel = (): BuilderPlugin => ({
  name: 'builder-plugin-babel',
  setup(api) {
    api.modifyBundlerChain(
      async (chain, { CHAIN_ID, isProd, getCompiledPath }) => {
        const config = api.getNormalizedConfig();
        if (!config.tools.babel) {
          // we would not use babel loader in rspack, unless user need to use.
          return;
        }

        const { applyUserBabelConfig } = await import('@modern-js/utils');

        const getBabelOptions = (config: NormalizedConfig) => {
          // 1. Create babel utils function about include/exclude,
          const includes: Array<string | RegExp> = [];
          const excludes: Array<string | RegExp> = [];

          const babelUtils = {
            addIncludes(items: string | RegExp | Array<string | RegExp>) {
              if (Array.isArray(items)) {
                includes.push(...items);
              } else {
                includes.push(items);
              }
            },
            addExcludes(items: string | RegExp | Array<string | RegExp>) {
              if (Array.isArray(items)) {
                excludes.push(...items);
              } else {
                excludes.push(items);
              }
            },
          };

          const baseConfig = {
            plugins: [],
            presets: [
              [
                require.resolve('@babel/preset-typescript'),
                DEFAULT_BABEL_PRESET_TYPESCRIPT_OPTIONS,
              ],
            ],
          };

          const userBabelConfig = applyUserBabelConfig(
            cloneDeep(baseConfig),
            config.tools.babel,
            babelUtils,
          );

          const notModify =
            isEqual(baseConfig, userBabelConfig) &&
            !includes?.length &&
            !excludes?.length;

          if (notModify) {
            return {};
          }

          const babelOptions: BabelOptions = {
            babelrc: false,
            configFile: false,
            compact: isProd,
            ...userBabelConfig,
          };

          return {
            babelOptions,
            includes,
            excludes,
          };
        };

        const {
          babelOptions,
          includes = [],
          excludes = [],
        } = getBabelOptions(config);

        if (!babelOptions) {
          return;
        }

        const rule = chain.module.rule(CHAIN_ID.RULE.JS);

        applyScriptCondition({
          rule,
          config,
          context: api.context,
          includes,
          excludes,
        });

        rule
          .test(mergeRegex(JS_REGEX, TS_REGEX))
          .use(CHAIN_ID.USE.BABEL)
          .loader(getCompiledPath('babel-loader'))
          .options(babelOptions);
      },
    );
  },
});
