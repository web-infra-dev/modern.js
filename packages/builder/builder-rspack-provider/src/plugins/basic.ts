import type { BuilderPlugin } from '../types';

/**
 * Provide some basic configs of rspack
 */
export const builderPluginBasic = (): BuilderPlugin => ({
  name: 'builder-plugin-basic',

  setup(api) {
    api.modifyRspackConfig(async (rspackConfig, { isProd }) => {
      // The base directory for resolving entry points and loaders from the configuration.
      rspackConfig.context = api.context.rootPath;

      rspackConfig.mode = isProd ? 'production' : 'development';

      rspackConfig.infrastructureLogging = {
        // Using `error` level to avoid `cache.PackFileCacheStrategy` logs
        level: 'error',
      };

      // todo: rspack not support configure
      // // if the chunk size exceeds 1MiB, we will throw a warning
      // chain.performance.maxAssetSize(1024 * 1024);
      // chain.performance.maxEntrypointSize(1024 * 1024);

      // // This will be futureDefaults in webpack 6
      // chain.module.parser.merge({
      //   javascript: {
      //     exportsPresence: 'error',
      //   },
      // });
    });
  },
});
