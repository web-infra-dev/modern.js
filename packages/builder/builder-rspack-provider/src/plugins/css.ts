import assert from 'assert';
import {
  isLooseCssModules,
  getBrowserslistWithDefault,
  isUseCssSourceMap,
  CSS_REGEX,
  type BuilderContext,
  BundlerChain,
  ModifyBundlerChainUtils,
} from '@modern-js/builder-shared';
import type { BuilderPlugin, NormalizedConfig } from '../types';
import type { AcceptedPlugin } from 'postcss';
import { isUseCssExtract, getCompiledPath } from '../shared';

type CssNanoOptions = {
  configFile?: string | undefined;
  preset?: [string, object] | string | undefined;
};

export const getCssnanoDefaultOptions = (): CssNanoOptions => ({
  preset: [
    'default',
    {
      // merge longhand will break safe-area-inset-top, so disable it
      // https://github.com/cssnano/cssnano/issues/803
      // https://github.com/cssnano/cssnano/issues/967
      mergeLonghand: false,
    },
  ],
});

export async function applyBaseCSSRule(
  rule: ReturnType<BundlerChain['module']['rule']>,
  config: NormalizedConfig,
  context: BuilderContext,
  { target, isProd, isServer, isWebWorker, CHAIN_ID }: ModifyBundlerChainUtils,
) {
  // 1. Check user config
  const enableExtractCSS = isUseCssExtract(config, target);
  const enableSourceMap = isUseCssSourceMap(config);
  // const enableCSSModuleTS = Boolean(config.output.enableCssModuleTSDeclaration);

  const { applyOptionsChain } = await import('@modern-js/utils');
  const browserslist = await getBrowserslistWithDefault(
    context.rootPath,
    config,
    target,
  );

  const localIdentName =
    config.output.cssModuleLocalIdentName ||
    // Using shorter classname in production to reduce bundle size
    (isProd ? '[hash:base64:5]' : '[path][name]__[local]--[hash:base64:5]');

  /**
   * postcss-module options.
   * different with css-loader, postcss-module without exportOnlyLocals option
   *
   * TODO: support css-module in rspack
   */
  const moduleConfig = {
    auto: config.output.disableCssModuleExtension ? isLooseCssModules : true,
    localsConvention: 'camelCase',
    generateScopedName: localIdentName,
  };

  const getPostcssConfig = () => {
    const extraPlugins: AcceptedPlugin[] = [];
    const utils = {
      addPlugins(plugins: AcceptedPlugin | AcceptedPlugin[]) {
        if (Array.isArray(plugins)) {
          extraPlugins.push(...plugins);
        } else {
          extraPlugins.push(plugins);
        }
      },
    };
    const enableCssMinify = !enableExtractCSS && isProd;

    const mergedConfig = applyOptionsChain(
      {
        postcssOptions: {
          plugins: [
            require(getCompiledPath('postcss-flexbugs-fixes')),
            require(getCompiledPath('postcss-custom-properties')),
            require(getCompiledPath('postcss-initial')),
            require(getCompiledPath('postcss-page-break')),
            require(getCompiledPath('postcss-font-variant')),
            require(getCompiledPath('postcss-media-minmax')),
            require(getCompiledPath('postcss-nesting')),
            require(getCompiledPath('autoprefixer'))(
              applyOptionsChain(
                {
                  flexbox: 'no-2009',
                  overrideBrowserslist: browserslist,
                },
                config.tools.autoprefixer,
              ),
            ),
            enableCssMinify
              ? require('cssnano')(getCssnanoDefaultOptions())
              : false,
          ].filter(Boolean),
        },
        sourceMap: enableSourceMap,
        modules: moduleConfig,
      },
      // postcss-loader will modify config
      config.tools.postcss || {},
      utils,
    );
    if (extraPlugins.length) {
      assert('postcssOptions' in mergedConfig);
      assert('plugins' in mergedConfig.postcssOptions!);
      mergedConfig.postcssOptions.plugins!.push(...extraPlugins);
    }
    return mergedConfig;
  };

  /**
   * TODO: support style-loader & ignore css (need rspack support inline css first)
   */
  //   if (isServer || isWebWorker) {
  //     const { default: ignoreCssLoader } = await import('../loaders/ignoreCssLoader');
  //     uses.push({
  //       name: CHAIN_ID.USE.IGNORE_CSS,
  //       loader: ignoreCssLoader,
  //     });
  //   }

  // TODO: use css-modules-typescript-loader
  // if (enableCSSModuleTS) {
  //   const { default: cssModulesTypescriptLoader } = await import(
  //     getCompiledPath('css-modules-typescript-loader')
  //   );

  //   uses.push({
  //     name: CHAIN_ID.USE.CSS_MODULES_TS,
  //     loader: cssModulesTypescriptLoader,
  //     options: {},
  //   });
  // }

  if (!isServer && !isWebWorker) {
    // todo: css module (exportOnlyLocals) required in server
    const postcssLoaderOptions = getPostcssConfig();

    rule
      .use(CHAIN_ID.USE.POSTCSS)
      .loader(require.resolve('@rspack/postcss-loader'))
      .options(postcssLoaderOptions)
      .end();
  }

  // todo: rspack not support sideEffects config
  // CSS imports should always be treated as sideEffects
  // rule.merge({ sideEffects: true });
}

export const builderPluginCss = (): BuilderPlugin => {
  return {
    name: 'builder-plugin-css',
    setup(api) {
      api.modifyBundlerChain(async (chain, utils) => {
        const config = api.getNormalizedConfig();

        const rule = chain.module.rule(utils.CHAIN_ID.RULE.CSS);
        rule.test(CSS_REGEX).type('css');

        await applyBaseCSSRule(rule, config, api.context, utils);
      });
    },
  };
};
