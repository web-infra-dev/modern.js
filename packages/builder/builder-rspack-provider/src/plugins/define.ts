import _ from '@modern-js/utils/lodash';
import type { BuilderPlugin, SourceConfig } from '../types';
import { setConfig } from '@modern-js/builder-shared';

export const builderPluginDefine = (): BuilderPlugin => ({
  name: 'builder-plugin-define',

  async setup(api) {
    api.modifyRspackConfig(async rspackConfig => {
      const config = api.getNormalizedConfig();

      const builtinVars: NonNullable<SourceConfig['globalVars']> = {
        'process.env.NODE_ENV': process.env.NODE_ENV || 'development',
      };

      // Serialize global vars. User can customize value of `builtinVars`.
      const globalVars = {
        ...builtinVars,
        ...(config.source.globalVars ?? {}),
      };

      const serializedVars = _.mapValues(globalVars, value =>
        JSON.stringify(value),
      );

      // Macro defines.
      const defineExprs = config.source.define ?? {};

      setConfig(rspackConfig, 'builtins.define', {
        ...(rspackConfig.builtins?.define ?? {}),
        ...serializedVars,
        ...defineExprs,
      });
    });
  },
});
