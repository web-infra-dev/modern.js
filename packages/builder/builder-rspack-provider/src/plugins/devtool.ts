import { isUseJsSourceMap } from '@modern-js/builder-shared';
import type { BuilderPlugin } from '../types';

export const PluginDevtool = (): BuilderPlugin => ({
  name: 'builder-plugin-devtool',

  setup(api) {
    api.modifyRspackConfig((rspackConfig, { isProd }) => {
      const config = api.getNormalizedConfig();

      if (!isUseJsSourceMap(config)) {
        rspackConfig.devtool = false;
      } else {
        const devtool = isProd
          ? // hide the source map URL in production to avoid Chrome warning
            'hidden-source-map'
          : 'cheap-module-source-map';
        rspackConfig.devtool = devtool;
      }
    });
  },
});
