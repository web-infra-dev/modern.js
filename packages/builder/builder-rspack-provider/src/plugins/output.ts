import {
  DEFAULT_PORT,
  addTrailingSlash,
  getDistPath,
  getFilename,
  type BuilderContext,
} from '@modern-js/builder-shared';
import type { BuilderPlugin, NormalizedConfig } from '../types';
import { isUseCssExtract } from '../shared';

function getPublicPath({
  config,
  isProd,
  context,
}: {
  config: NormalizedConfig;
  isProd: boolean;
  context: BuilderContext;
}) {
  const { dev, output } = config;

  let publicPath = '/';

  if (isProd) {
    if (output.assetPrefix) {
      publicPath = output.assetPrefix;
    }
  } else if (typeof dev.assetPrefix === 'string') {
    publicPath = dev.assetPrefix;
  } else if (dev.assetPrefix === true) {
    const hostname = context.devServer?.hostname || 'localhost';
    const port = context.devServer?.port || DEFAULT_PORT;
    publicPath = `//${hostname}:${port}/`;
  }

  return addTrailingSlash(publicPath);
}

export const builderPluginOutput = (): BuilderPlugin => ({
  name: 'builder-plugin-output',

  setup(api) {
    api.modifyRspackConfig(
      async (rspackConfig, { isProd, isServer, target }) => {
        const config = api.getNormalizedConfig();
        const jsPath = getDistPath(config.output, 'js');

        const publicPath = getPublicPath({
          config,
          isProd,
          context: api.context,
        });

        // js output
        const jsFilename = getFilename(config.output, 'js', isProd);

        rspackConfig.output = {
          path: api.context.distPath,
          filename: `${jsPath}/${jsFilename}`,
          chunkFilename: `${jsPath}/async/${jsFilename}`,
          publicPath,
          hashFunction: 'xxhash64',
        };

        // css output
        if (isUseCssExtract(config, target)) {
          const cssPath = getDistPath(config.output, 'css');
          const cssFilename = getFilename(config.output, 'css', isProd);

          rspackConfig.output.cssFilename = `${cssPath}/${cssFilename}`;
          rspackConfig.output.cssChunkFilename = `${cssPath}/async/${cssFilename}`;
        }

        // server output
        if (isServer) {
          const serverPath = getDistPath(config.output, 'server');
          const filename = `${serverPath}/[name].js`;

          rspackConfig.output = {
            path: api.context.distPath,
            filename,
            chunkFilename: filename,
            // todo
            // libraryTarget: 'commonjs2',
            publicPath,
            hashFunction: 'xxhash64',
          };
        }
      },
    );
  },
});
