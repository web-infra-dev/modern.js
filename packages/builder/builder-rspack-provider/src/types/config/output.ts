import type {
  SharedOutputConfig,
  NormalizedSharedOutputConfig,
  RemOptions,
} from '@modern-js/builder-shared';

export type OutputConfig = Omit<SharedOutputConfig, 'legalComments'> & {
  /**
   * Copies the specified file or directory to the dist directory.
   * TODO: not support yet
   */
  // copy?: CopyPluginOptions | CopyPluginOptions['patterns'];
  /**
   * Convert px to rem in CSS.
   */
  convertToRem?: boolean | RemOptions;
  /**
   * At build time, prevent some `import` dependencies from being packed into bundles in your code, and instead fetch them externally at runtime.
   */
  externals?: Record<string, string>;
};

export type NormalizedOutputConfig = OutputConfig &
  NormalizedSharedOutputConfig;
