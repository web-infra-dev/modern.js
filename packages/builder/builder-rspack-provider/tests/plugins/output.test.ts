import { expect, describe, it } from 'vitest';
import { createBuilder } from '../helper';
import { builderPluginOutput } from '@/plugins/output';

describe('plugins/output', () => {
  it('should set output correctly', async () => {
    const builder = await createBuilder({
      plugins: [builderPluginOutput()],
    });

    const {
      origin: { bundlerConfigs },
    } = await builder.inspectConfig();
    expect(bundlerConfigs[0]).toMatchSnapshot();
  });

  it('should allow to custom server directory with distPath.server', async () => {
    const builder = await createBuilder({
      plugins: [builderPluginOutput()],
      target: ['node'],
      builderConfig: {
        output: {
          distPath: {
            server: 'server',
          },
        },
      },
    });

    const {
      origin: { bundlerConfigs },
    } = await builder.inspectConfig();
    expect(bundlerConfigs[0]).toMatchSnapshot();
  });

  it('should allow to set distPath.js and distPath.css to empty string', async () => {
    const builder = await createBuilder({
      plugins: [builderPluginOutput()],
      builderConfig: {
        output: {
          distPath: {
            js: '',
            css: '',
          },
        },
      },
    });

    const {
      origin: { bundlerConfigs },
    } = await builder.inspectConfig();
    expect(bundlerConfigs[0]).toMatchSnapshot();
  });

  it('should allow to use filename.js to modify filename', async () => {
    const builder = await createBuilder({
      plugins: [builderPluginOutput()],
      builderConfig: {
        output: {
          filename: {
            js: 'foo.js',
            css: '[name].css',
          },
        },
      },
    });

    const {
      origin: { bundlerConfigs },
    } = await builder.inspectConfig();
    expect(bundlerConfigs[0]).toMatchSnapshot();
  });

  it('should allow to use copy plugin', async () => {
    const builder = await createBuilder({
      plugins: [builderPluginOutput()],
      builderConfig: {
        output: {
          copy: {
            patterns: [
              {
                from: 'test',
              },
            ],
          },
        },
      },
    });

    const {
      origin: { bundlerConfigs },
    } = await builder.inspectConfig();
    expect(bundlerConfigs[0]).toMatchSnapshot();
  });

  it('should allow to use copy plugin with multiply config', async () => {
    const builder = await createBuilder({
      plugins: [builderPluginOutput()],
      builderConfig: {
        output: {
          copy: [
            {
              from: 'test',
            },
            'src/assets/',
          ],
        },
        tools: {
          bundlerChain: (chain, { CHAIN_ID }) => {
            chain.plugin(CHAIN_ID.PLUGIN.COPY).tap(args => [
              {
                patterns: [...(args[0]?.patterns || []), 'tests/'],
              },
            ]);
          },
        },
      },
    });

    const {
      origin: { bundlerConfigs },
    } = await builder.inspectConfig();
    expect(bundlerConfigs[0]).toMatchSnapshot();
  });
});
