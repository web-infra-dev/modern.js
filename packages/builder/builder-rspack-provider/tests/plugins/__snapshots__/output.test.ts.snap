// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugins/output > should allow to custom server directory with distPath.server 1`] = `
{
  "output": {
    "chunkFilename": "server/[name].js",
    "cssChunkFilename": "static/css/async/[name].css",
    "cssFilename": "static/css/[name].css",
    "filename": "server/[name].js",
    "hashFunction": "xxhash64",
    "libraryTarget": "commonjs2",
    "path": "<ROOT>/dist",
    "pathinfo": false,
    "publicPath": "/",
  },
}
`;

exports[`plugins/output > should allow to set distPath.js and distPath.css to empty string 1`] = `
{
  "output": {
    "chunkFilename": "async/[name].js",
    "cssChunkFilename": "async/[name].css",
    "cssFilename": "[name].css",
    "filename": "[name].js",
    "hashFunction": "xxhash64",
    "path": "<ROOT>/dist",
    "pathinfo": false,
    "publicPath": "/",
  },
}
`;

exports[`plugins/output > should allow to use copy plugin 1`] = `
{
  "output": {
    "chunkFilename": "static/js/async/[name].js",
    "cssChunkFilename": "static/css/async/[name].css",
    "cssFilename": "static/css/[name].css",
    "filename": "static/js/[name].js",
    "hashFunction": "xxhash64",
    "path": "<ROOT>/dist",
    "pathinfo": false,
    "publicPath": "/",
  },
  "plugins": [
    Plugin {
      "_options": {
        "patterns": [
          {
            "force": false,
            "from": "test",
            "globOptions": {},
            "noErrorOnMissing": false,
            "priority": 0,
          },
        ],
      },
      "name": "CopyRspackPlugin",
    },
  ],
}
`;

exports[`plugins/output > should allow to use copy plugin with multiply config 1`] = `
{
  "output": {
    "chunkFilename": "static/js/async/[name].js",
    "cssChunkFilename": "static/css/async/[name].css",
    "cssFilename": "static/css/[name].css",
    "filename": "static/js/[name].js",
    "hashFunction": "xxhash64",
    "path": "<ROOT>/dist",
    "pathinfo": false,
    "publicPath": "/",
  },
  "plugins": [
    Plugin {
      "_options": {
        "patterns": [
          {
            "force": false,
            "from": "test",
            "globOptions": {},
            "noErrorOnMissing": false,
            "priority": 0,
          },
          {
            "force": false,
            "from": "src/assets/",
            "globOptions": {},
            "noErrorOnMissing": false,
            "priority": 0,
          },
          {
            "force": false,
            "from": "tests/",
            "globOptions": {},
            "noErrorOnMissing": false,
            "priority": 0,
          },
        ],
      },
      "name": "CopyRspackPlugin",
    },
  ],
}
`;

exports[`plugins/output > should allow to use filename.js to modify filename 1`] = `
{
  "output": {
    "chunkFilename": "static/js/async/foo.js",
    "cssChunkFilename": "static/css/async/[name].css",
    "cssFilename": "static/css/[name].css",
    "filename": "static/js/foo.js",
    "hashFunction": "xxhash64",
    "path": "<ROOT>/dist",
    "pathinfo": false,
    "publicPath": "/",
  },
}
`;

exports[`plugins/output > should set output correctly 1`] = `
{
  "output": {
    "chunkFilename": "static/js/async/[name].js",
    "cssChunkFilename": "static/css/async/[name].css",
    "cssFilename": "static/css/[name].css",
    "filename": "static/js/[name].js",
    "hashFunction": "xxhash64",
    "path": "<ROOT>/dist",
    "pathinfo": false,
    "publicPath": "/",
  },
}
`;
